driver/api/impl/impl.cpp

```diff
@@ -1166,74 +1264,212 @@ SQLRETURN fillBinding(
     return result_set.extractField(row_idx, column_idx, binding_info);
 }
 
-SQLRETURN GetData(
-    Statement & statement,
-    SQLUSMALLINT column_or_param_number,
-    const BindingInfo & binding_info
-) {
-    if (!statement.hasResultSet())
-        throw SqlException("Column info is not available", "07005");
-
-    auto & result_set = statement.getResultSet();
-
-    if (result_set.getCurrentRowPosition() < 1)
-        throw SqlException("Invalid cursor state", "24000");
-
-    if (column_or_param_number < 1)
-        throw SqlException("Invalid descriptor index", "07009");
-
-    const auto column_idx = column_or_param_number - 1;
-    const auto row_idx = result_set.getCurrentRowPosition() - result_set.getCurrentRowSetPosition();
-
-    return fillBinding(statement, result_set, row_idx, column_idx, binding_info);
-}
-
-SQLRETURN FetchScroll(
+SQLRETURN fetchBindings(
     Statement & statement,
     SQLSMALLINT orientation,
     SQLLEN offset
 ) {
-    const auto row_set_size = statement.getEffectiveDescriptor(SQL_ATTR_APP_ROW_DESC).getAttrAs<SQLULEN>(SQL_DESC_ARRAY_SIZE, 1);
-    auto * rows_fetched_ptr = statement.getEffectiveDescriptor(SQL_ATTR_IMP_ROW_DESC).getAttrAs<SQLULEN *>(SQL_DESC_ROWS_PROCESSED_PTR, 0);
+    auto & ard_desc = statement.getEffectiveDescriptor(SQL_ATTR_APP_ROW_DESC);
+    auto & ird_desc = statement.getEffectiveDescriptor(SQL_ATTR_IMP_ROW_DESC);
+
+    const auto row_set_size = ard_desc.getAttrAs<SQLULEN>(SQL_DESC_ARRAY_SIZE, 1);
+    auto * rows_fetched_ptr = ird_desc.getAttrAs<SQLULEN *>(SQL_DESC_ROWS_PROCESSED_PTR, 0);
+    auto * array_status_ptr = ird_desc.getAttrAs<SQLUSMALLINT *>(SQL_DESC_ARRAY_STATUS_PTR, 0);
 
     if (rows_fetched_ptr)
         *rows_fetched_ptr = 0;
 
-    if (!statement.hasResultSet())
+    if (!statement.hasResultSet() || row_set_size == 0) {
+        if (array_status_ptr) {
+            for (std::size_t row_idx = 0; row_idx < row_set_size; ++row_idx) {
+                array_status_ptr[row_idx] = SQL_ROW_NOROW;
+            }
+        }
+
         return SQL_NO_DATA;
+    }
 
     auto & result_set = statement.getResultSet();
-
     const auto rows_fetched = result_set.fetchRowSet(orientation, offset, row_set_size);
 
     if (rows_fetched == 0) {
         statement.getDiagHeader().setAttr(SQL_DIAG_ROW_COUNT, result_set.getAffectedRowCount());
+
+        if (array_status_ptr) {
+            for (std::size_t row_idx = 0; row_idx < row_set_size; ++row_idx) {
+                array_status_ptr[row_idx] = SQL_ROW_NOROW;
+            }
+        }
+
         return SQL_NO_DATA;
     }
 
     if (rows_fetched_ptr)
         *rows_fetched_ptr = rows_fetched;
 
-    auto res = SQL_SUCCESS;
+    const auto ard_record_count = ard_desc.getRecordCount();
```

Comment:
can count and container somehow diverge ?

Reply:
No, that is ensured exactly by the next line.